okhttp/src/main/java/okhttp3/internal/connection/RealConnectionPool.kt

```diff
@@ -14,249 +14,270 @@
+    var i = 0
+    while (i < references.size) {
+      val reference = references[i]
 
       if (reference.get() != null) {
-        i++;
-        continue;
+        i++
+        continue
       }
 
       // We've discovered a leaked transmitter. This is an application bug.
-      TransmitterReference transmitterRef = (TransmitterReference) reference;
-      String message = "A connection to " + connection.route().address().url()
-          + " was leaked. Did you forget to close a response body?";
-      Platform.get().logCloseableLeak(message, transmitterRef.callStackTrace);
+      val transmitterRef = reference as TransmitterReference
+      val message = "A connection to ${connection.route().address().url()} was leaked. " +
+          "Did you forget to close a response body?"
+      Platform.get().logCloseableLeak(message, transmitterRef.callStackTrace)
 
-      references.remove(i);
-      connection.noNewExchanges = true;
+      references.removeAt(i)
+      connection.noNewExchanges = true
 
       // If this was the last allocation, the connection is eligible for immediate eviction.
       if (references.isEmpty()) {
-        connection.idleAtNanos = now - keepAliveDurationNs;
-        return 0;
+        connection.idleAtNanos = now - keepAliveDurationNs
+        return 0
       }
     }
 
-    return references.size();
+    return references.size
   }
 
-  /** Track a bad route in the route database. Other routes will be attempted first. */
-  public void connectFailed(Route failedRoute, IOException failure) {
+  /** Track a bad route in the route database. Other routes will be attempted first.  */
+  fun connectFailed(failedRoute: Route, failure: IOException) {
     // Tell the proxy selector when we fail to connect on a fresh connection.
     if (failedRoute.proxy().type() != Proxy.Type.DIRECT) {
-      Address address = failedRoute.address();
+      val address = failedRoute.address()
       address.proxySelector().connectFailed(
-          address.url().uri(), failedRoute.proxy().address(), failure);
+          address.url().uri(), failedRoute.proxy().address(), failure)
     }
 
-    routeDatabase.failed(failedRoute);
+    routeDatabase.failed(failedRoute)
+  }
+
+  companion object {
+    /**
+     * Background threads are used to cleanup expired connections. There will be at most a single
+     * thread running per connection pool. The thread pool executor permits the pool itself to be
+     * garbage collected.
+     */
+    private val executor = ThreadPoolExecutor(
+        0, // corePoolSize.
+        Integer.MAX_VALUE, // maximumPoolSize.
+        60L, TimeUnit.SECONDS,  // keepAliveTime.
+        SynchronousQueue(),
+        Util.threadFactory("OkHttp ConnectionPool", true)
+    )
+  }
+}
+
+/**
+ * Lock and wait a duration in nanoseconds. Unlike [java.lang.Object.wait] this interprets 0 as
+ * "don't wait" instead of "wait forever".
+ */
+@Throws(InterruptedException::class)
+private fun Any.waitNanos(nanos: Long) {
```

Comment:
Where should I stash this?

Reply:
Here seems fine unless you're going to need it in more than one place.